src/Psl/Shell/execute.php

```diff
@@ -0,0 +1,89 @@
+use Psl\Vec;
+
+use function fclose;
+use function is_dir;
+use function is_resource;
+use function proc_close;
+use function proc_open;
+use function stream_get_contents;
+
+/**
+ * Execute an external program.
+ *
+ * @param string                $command            The command to execute.
+ * @param list<string>          $arguments          The command arguments listed as separate entries.
+ * @param string                $working_directory  The initial working directory for the command.
+ *                                                  This must be an absolute directory path, or null
+ *                                                  if you want to use the default value (the working dir of the
+ *                                                  current directory )
+ * @param array<string, string> $environment        An array with the environment variables for the command that
+ *                                                  will be run.
+ * @param bool                  $escape_arguments   If set to true ( default ), all $arguments will be escaped using
+ *                                                  `Shell\escape_argument`.
+ *
+ * @psalm-taint-sink shell $command
+ *
+ * @throws Exception\FailedExecutionException   In case the command resulted in an exit code other than 0.
+ * @throws Exception\PossibleAttackException    In case the command being run is suspicious ( e.g: contains NULL byte ).
+ * @throws Exception\RuntimeException           In case $working_directory doesn't exist, or unable to create a new
+ *                                              process.
+ */
+function execute(
+    string $command,
+    array $arguments = [],
+    ?string $working_directory = null,
+    array $environment = [],
+    bool $escape_arguments = true
+): string {
+    if ($escape_arguments) {
+        $arguments = Vec\map($arguments, static fn(string $argument): string => escape_argument($argument));
+    }
+
+    $commandline = Str\join([$command, ...$arguments], ' ');
+
+    /** @psalm-suppress MissingThrowsDocblock - safe ( $offset is within-of-bounds ) */
+    if (Str\contains($commandline, "\0")) {
+        throw new Exception\PossibleAttackException('NULL byte detected.');
+    }
+
+    $descriptor = [
+        1 => ['pipe', 'w'],
+        2 => ['pipe', 'w'],
+    ];
+
+    $environment = Dict\merge(Env\get_vars(), $environment);
+    $working_directory = $working_directory ?? Env\current_dir();
+    if (!is_dir($working_directory)) {
+        throw new Exception\RuntimeException('$working_directory does not exist.');
+    }
+
+    $process = proc_open($commandline, $descriptor, $pipes, $working_directory, $environment);
+    // @codeCoverageIgnoreStart
+    // not sure how to replicate this, but it can happen \_o.o_/
+    if (!is_resource($process)) {
+        throw new Exception\RuntimeException('Failed to open a new process.');
+    }
+    // @codeCoverageIgnoreEnd
+
+    $stdout_content = stream_get_contents($pipes[1]);
```

Comment:
Here you are fetching stdout and stderr. However, there are tools that report to both stdout and stderr.
For example take "composer run".
It reports the command it is executing to stdout, but the results to stdin.
If the run fails, it reports back to stdout.

By using `stream_get_contents`, you are blocking process execution untill the stream is closed (or eof).
This means that there is no way to combine stdout and stderr back to the actual output you see when you run the command on the screen.

So what would be better? Maybe wrap the output in a result and keep the target output linked to the actually outputted content? Something like:

```php
$output = [];
$output[] = ["STDERR", 'Starting command composer run mytask'];
$output[] = ["STDOUT", 'the actual output of the command'];
$output[] = ["STDERR", 'Execution task mytask failed : some context'];
```


You can then wait for the process exection like this:

```php
do {
    $status = proc_get_status($process);
    // Check if content is added to stdout / stderr and link it to the custom output class.
} while ($status && $status['running']);
```

That would give you a chronological order of messages, filterable on what stream it was added.

WDYT?

Reply:
that's why i said this is not a replacement for `proc_*`, it is more of a replacement to `exec` and `shell_exec` but it allows you to change environment variables, working_directory, escape things for you by default .. etc.

we will have something in the future that can totally replace `proc_*`, but it's not the goal of `Shell\execute`